File: android/app/src/main/kotlin/com/goalmaker/app/domain/design/DesignTokens.kt
package com.goalmaker.app.domain.design

import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.boolean
import kotlinx.serialization.json.float
import kotlinx.serialization.json.int
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive

/**
 * Everything contracts/design/themes.json defines, as the phone uses it (ADR 0008). The file is
 * checked by tools/check_design_tokens.py, so parsing trusts its shape and fails loudly otherwise.
 */
class DesignTokens(
    val defaultThemeId: String,
    val themes: List<ThemeDefinition>,
    val areaColors: List<AreaColor>,
    val density: Density,
    val motion: MotionTokens,
) {
    private val byId = themes.associateBy(ThemeDefinition::id)

    /** The theme with [id], or the default when it's unknown (for example after a theme is retired). */
    fun theme(id: String?): ThemeDefinition = byId[id] ?: byId.getValue(defaultThemeId)

    fun areaColor(id: String): AreaColor? = areaColors.firstOrNull { it.id == id }

    companion object {
        fun parse(json: String): DesignTokens {
            val root = Json.parseToJsonElement(json).jsonObject
            val themes = root.obj("themes").map { (id, element) -> theme(id, element.jsonObject) }
            val areas = root.obj("areaPalette").getValue("colors").jsonArray.map { element ->
                val area = element.jsonObject
                AreaColor(
                    id = area.text("id"),
                    swatch = parseColor(area.text("swatch")),
                    light = chip(area.obj("light")),
                    dark = chip(area.obj("dark")),
                )
            }
            val phone = root.obj("density").obj("phone")
            val motion = root.obj("motion")
            return DesignTokens(
                defaultThemeId = root.text("defaultTheme"),
                themes = themes,
                areaColors = areas,
                density = Density(phone.int("rowMinHeight"), phone.int("rowGap"), phone.int("pagePadding"), phone.int("cardPadding")),
                motion = MotionTokens(motion.int("quick"), motion.int("standard"), motion.int("emphasized")),
            )
        }

        private fun theme(id: String, theme: JsonObject): ThemeDefinition {
            val typography = theme.obj("typography")
            val shape = theme.obj("shape")
            val dark = theme.obj("dark")
            return ThemeDefinition(
                id = id,
                name = theme.text("name"),
                summary = theme.text("summary"),
                typography = ThemeTypography(
                    heading = typeStyle(typography.obj("heading")),
                    body = typography.obj("body").let { body ->
                        BodyStyle(body.text("family"), body.int("weight"), body.int("strongWeight"), body.float("width"))
                    },
                    number = typeStyle(typography.obj("number")),
                ),
                shapes = ThemeShapes(shape.int("card"), shape.int("row"), shape.int("checkbox"), shape.int("button")),
                light = palette(theme.obj("light")),
                dark = palette(dark),
                black = palette(JsonObject(dark + theme.obj("black"))),
                logo = theme.obj("logo").let { logo -> LogoColors(logo.color("tile"), logo.color("letter"), logo.color("arrow")) },
            )
        }

        private fun typeStyle(style: JsonObject) = TypeStyle(
            family = style.text("family"),
            weight = style.int("weight"),
            width = style.float("width"),
            italic = style["italic"]?.jsonPrimitive?.boolean ?: false,
            uppercase = style["uppercase"]?.jsonPrimitive?.boolean ?: false,
            tracking = style["tracking"]?.jsonPrimitive?.float ?: 0f,
        )

        private fun palette(roles: JsonObject) = Palette(
            background = roles.color("background"),
            surface = roles.color("surface"),
            surfaceVariant = roles.color("surfaceVariant"),
            text = roles.color("text"),
            textMuted = roles.color("textMuted"),
            outline = roles.color("outline"),
            primary = roles.color("primary"),
            onPrimary = roles.color("onPrimary"),
            accent = roles.color("accent"),
            onAccent = roles.color("onAccent"),
            hero = roles.color("hero"),
            onHero = roles.color("onHero"),
            heroAccent = roles.color("heroAccent"),
            danger = roles.color("danger"),
        )

        private fun chip(pair: JsonObject) = ChipColors(pair.color("container"), pair.color("content"))

        /** "#RRGGBB" as an opaque ARGB integer. */
        fun parseColor(hex: String): Int {
            require(hex.length == 7 && hex[0] == '#') { "Not a #RRGGBB color: $hex" }
            return (0xFF000000 or hex.substring(1).toLong(16)).toInt()
        }

        private fun JsonObject.obj(name: String) = getValue(name).jsonObject

        private fun JsonObject.text(name: String) = getValue(name).jsonPrimitive.content

        private fun JsonObject.int(name: String) = getValue(name).jsonPrimitive.int

        private fun JsonObject.float(name: String) = getValue(name).jsonPrimitive.float

        private fun JsonObject.color(name: String): Int = parseColor(text(name))
    }
}

File: windows/src/GoalMaker.Core/Design/DesignTokens.cs
using System.Globalization;
using System.Text.Json;

namespace GoalMaker.Core.Design;

/// <summary>
/// Everything contracts/design/themes.json defines, as the Windows app uses it (ADR 0008). The file is
/// checked by tools/check_design_tokens.py, so loading trusts its shape and fails loudly otherwise.
/// </summary>
public sealed class DesignTokens
{
    private readonly Dictionary<string, ThemeDefinition> byId;

    public DesignTokens(string defaultThemeId, IReadOnlyList<ThemeDefinition> themes, IReadOnlyList<AreaColor> areaColors, Density density, MotionTokens motion)
    {
        DefaultThemeId = defaultThemeId;
        Themes = themes;
        AreaColors = areaColors;
        Density = density;
        Motion = motion;
        byId = themes.ToDictionary(theme => theme.Id, StringComparer.Ordinal);
    }

    public string DefaultThemeId { get; }

    public IReadOnlyList<ThemeDefinition> Themes { get; }

    public IReadOnlyList<AreaColor> AreaColors { get; }

    public Density Density { get; }

    public MotionTokens Motion { get; }

    /// <summary>The theme with <paramref name="id"/>, or the default when it's unknown (for example a retired theme).</summary>
    public ThemeDefinition Theme(string? id) =>
        id is not null && byId.TryGetValue(id, out var theme) ? theme : byId[DefaultThemeId];

    public AreaColor? AreaColor(string id) => AreaColors.FirstOrDefault(area => area.Id == id);

    public static DesignTokens Load(Stream json)
    {
        using var document = JsonDocument.Parse(json);
        var root = document.RootElement;
        var themes = root.GetProperty("themes").EnumerateObject().Select(theme => Theme(theme.Name, theme.Value)).ToList();
        var areas = root.GetProperty("areaPalette").GetProperty("colors").EnumerateArray()
            .Select(area => new AreaColor(
                area.GetProperty("id").GetString()!,
                ParseColor(area.GetProperty("swatch").GetString()!),
                Chip(area.GetProperty("light")),
                Chip(area.GetProperty("dark"))))
            .ToList();
        var windows = root.GetProperty("density").GetProperty("windows");
        var motion = root.GetProperty("motion");
        return new DesignTokens(
            root.GetProperty("defaultTheme").GetString()!,
            themes,
            areas,
            new Density(
                windows.GetProperty("rowMinHeight").GetInt32(),
                windows.GetProperty("rowGap").GetInt32(),
                windows.GetProperty("pagePadding").GetInt32(),
                windows.GetProperty("cardPadding").GetInt32()),
            new MotionTokens(
                motion.GetProperty("quick").GetInt32(),
                motion.GetProperty("standard").GetInt32(),
                motion.GetProperty("emphasized").GetInt32()));
    }

    /// <summary>"#RRGGBB" as an opaque ARGB value.</summary>
    public static uint ParseColor(string hex)
    {
        if (hex.Length != 7 || hex[0] != '#')
        {
            throw new FormatException($"Not a #RRGGBB color: {hex}");
        }

        return 0xFF000000u | uint.Parse(hex.AsSpan(1), NumberStyles.HexNumber, CultureInfo.InvariantCulture);
    }

    private static ThemeDefinition Theme(string id, JsonElement theme)
    {
        var typography = theme.GetProperty("typography");
        var shape = theme.GetProperty("shape");
        var dark = theme.GetProperty("dark");
        var black = theme.GetProperty("black");
        var body = typography.GetProperty("body");
        var logo = theme.GetProperty("logo");
        return new ThemeDefinition(
            id,
            theme.GetProperty("name").GetString()!,
            theme.GetProperty("summary").GetString()!,
            new ThemeTypography(
                Style(typography.GetProperty("heading")),
                new BodyStyle(
                    body.GetProperty("family").GetString()!,
                    body.GetProperty("weight").GetInt32(),
                    body.GetProperty("strongWeight").GetInt32(),
                    body.GetProperty("width").GetDouble()),
                Style(typography.GetProperty("number"))),
            new ThemeShapes(
                shape.GetProperty("card").GetInt32(),
                shape.GetProperty("row").GetInt32(),
                shape.GetProperty("checkbox").GetInt32(),
                shape.GetProperty("button").GetInt32()),
            Palette(role => theme.GetProperty("light").GetProperty(role).GetString()!),
            Palette(role => dark.GetProperty(role).GetString()!),
            Palette(role => (black.TryGetProperty(role, out var over) ? over : dark.GetProperty(role)).GetString()!),
            new LogoColors(
                ParseColor(logo.GetProperty("tile").GetString()!),
                ParseColor(logo.GetProperty("letter").GetString()!),
                ParseColor(logo.GetProperty("arrow").GetString()!)));
    }

    private static TypeStyle Style(JsonElement style) => new(
        style.GetProperty("family").GetString()!,
        style.GetProperty("weight").GetInt32(),
        style.GetProperty("width").GetDouble(),
        style.TryGetProperty("italic", out var italic) && italic.GetBoolean(),
        style.TryGetProperty("uppercase", out var uppercase) && uppercase.GetBoolean(),
        style.TryGetProperty("tracking", out var tracking) ? tracking.GetDouble() : 0);

    private static Palette Palette(Func<string, string> role) => new(
        ParseColor(role("background")),
        ParseColor(role("surface")),
        ParseColor(role("surfaceVariant")),
        ParseColor(role("text")),
        ParseColor(role("textMuted")),
        ParseColor(role("outline")),
        ParseColor(role("primary")),
        ParseColor(role("onPrimary")),
        ParseColor(role("accent")),
        ParseColor(role("onAccent")),
        ParseColor(role("hero")),
        ParseColor(role("onHero")),
        ParseColor(role("heroAccent")),
        ParseColor(role("danger")));

    private static ChipColors Chip(JsonElement pair) =>
        new(ParseColor(pair.GetProperty("container").GetString()!), ParseColor(pair.GetProperty("content").GetString()!));
}
